transit.md

```diff
@@ -167,23 +170,164 @@ The Transit client can attempt connections to multiple relays, and uses the
+peer. This way either side can suggest the relays to use. The connection hints
 provided by the Transit instance include the locally-configured relay, along
 with the dynamically-determined direct hints. Both should be delivered to the
 peer.
 
+
+## Canonical abilities encodings
+
+The transit protocol relies on an existing secured (but possibly low-bandwidth)
+communication channel to exchange the abilities and hints. Thus, it has no
+influence over how they are encoded. However, we make a suggestion using JSON
+messages for other protocols to use.
+
+Abilities are encoded as a list, each item having a `type` tag. An ability may
+not appear more than once in the list. Setting `relay-v2` requires also setting
+`relay-v2` for backwards compatibility.
+
+```json
+[
+  {
+    "type": "<string, one of {direct-tcp-v1, relay-v1, tor-tcp-v1}>"
+  }
+]
+```
+
+Example for the full abilities set:
+
+```json
+[
+  { "type": "direct-tcp-v1" },
+  { "type": "relay-v1" },
+  { "type": "relay-v2" },
+  { "type": "tor-tcp-v1" }
+]
+```
+
+## Canonical hint encodings
+
+Hints are encoded as list of objects. Every object contains a `type` field,
+which further determines its encoding.
+
+### `direct-tcp-v1`, `tor-tcp-v1`
+
+```json
+{
+  "type": "direct-tcp-v1" or "tor-tcp-v1",
+  "hostname": <string>,
+  "port": <u16>,
+  "priority": <float; optional>
+}
+```
+
+`hostname` must be compliant with the `host` part of an URL, i.e. it may be an
+IP address or a domain. Furthermore, IPv6 link-local addresses are not supported.
+
+### `relay-v1`, `relay-v2`
+
+```json
+{
+  "type": "relay-v1",
+  "hints": [
+    {
+      "hostname": "<string>",
+      "port": "<tcp port>",
+      "priority": "<number, usually [0..1], optional>"
+    },
+    …
+  ],
+}
+```
+
+```json
+{
+  "type": "relay-v2",
+  "name": "<string, optional>",
+  "hints": [
+    {
+      "type": "tcp",
+      "hostname": "<string>",
+      "port": "<tcp port>"
+    },
+    {
+      "type": "websockets",
+      "url": "<url>"
+    },
+    …
+  ],
+}
+```
+
+A relay server may be reachable at multiple different end points and using
+different protocols. All hinted end points for a relay server must point to the
+same location, and the relay server must be able to connect any two of these
+endpoints. If this is not the case, the relay server must be advertized using
+multiple distinct hints instead (one per endpoint). Furthermore, a relay server
+may be given a human readable name for UI purposes.
+
+Hints have a `type` field, of which the currently known values are `tcp` and
+`websockets`. Hints of unknown type must be ignored. The `type` field is only
+required for `relay-v2` encoding. In the `relay-v1` incoding, only `tcp` hints
+are supported and the `type` field is implicit.
+
+The encoding of a `tcp` hint is equivalent to the one used by `direct-tcp-v1`.
+A hint of type `websockets` has an `url` field instead, which points to the
+WebSocket. Both `ws://` and `wss://` protocols must be supported.
```

Comment:
Why _must_ unencrypted connections be supported? (I understand that there's encryption already so using `wss` is maybe redundant, but also browsers or other restricted environments may not be allowed to make non-`wss` connections .. so tempted to change this to something like:

> The `wss://` protocol must be supported and we recommend supporting `ws://` as well whenever possible (the underlying messages are already encrypted by default).

Reply:
That's a good question, I obviously didn't think it through. The question is, *who* must *what*: we may either put requirements on the client side or on the relay side; the requirements for the others follow from that.

My initial idea was that clients must support both since a relay may not provide `wss` (this does not really hurt since all libraries should support it; it's more a question of policy). Alternatively, we could require all relays to support both. Then, it would not matter which one is used in the URL and it would be up to the client which one to use.

Reply:
A client in a Web browser won't be able to support `ws://` in general.

Reply:
Wait, what? Do they already enforce TLS?

Reply:
I thought on this some more, and made two observations:

- Unlike for the rendezvous server, TLS does not change anything to our security model
- Even if it did, hosting a relay server for a private network would be just fine with or without TLS

Thus, I propose the following:

- If the Relay server supports TLS, it should be advertized as `wss`
- `wss` implies `ws`, i.e. insecure WebSockets connections must always be accepted
- The clients are free in their choice, but encouraged to support both.

Reply:
For Web browsers: it _might_ be possible to think of situations where you'd be able to do unencrypted connections, but the page itself should be served by TLS which as far as I understand means you can't make unencrypted WebSocket connections. (I'm not 100% sure about this, browsers are fickle beasts ;) but [MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_client_applications) says, "This should use the URL scheme wss://, although some software may allow you to use the insecure ws:// for local connections. "

I guess partially I'm just reluctant to say "you _must_ support unencrypted connections" these days: that would mean someone deploying this has to understand enough about the protocols etc to learn that the underlying data packets are ("probably" -- that seems to be optional from the wording?) encrypted already. I'm presuming devops practice these days is normalizing on "everything is transport-encrypted" -- that's why I wanted to leave room in the spec. wording for someone to choose "everything is encrypted, always, and we don't allow unencrypted protocols".

(That is, an implementer could easily conclude from "must support both" that even having an option to turn one or the other off is bad).

Reply:
Oh I didn't know this, but it makes sense. How about this then:

- Both relay servers and clients *should* support `wss`, and it should be preferred over `ws`.
- If a relay server supports both, advertise that endpoint using two hints.

Reply:
yeah that sounds good